contrib/document_constants.py

```diff
@@ -0,0 +1,297 @@
+    '| %-4s | %-8s | %-12s |'
+    """
+    fmt = "|"
+    for length in lengths:
+        fmt += f" %-{length}s |"
+    return fmt
+
+
+def create_table_header_separator(lengths: Sequence[str]) -> str:
+    """Returns the Markdown table separator the columns:
+
+    >>> create_table_header_separator([4, 8])
+    '|:-----|:---------|'
+    >>> create_table_header_separator([4, 8, 12])
+    '|:-----|:---------|:-------------|'
+    """
+    fmt = "|"
+    for length in lengths:
+        dashes = '-' * length
+        fmt += f":{dashes}-|"
+    return fmt
+
+
+def create_table(
+    header: Sequence[str],
+    rows: Sequence[Sequence[str]],
+    size_rounding: int = 4,
+    omit_empty_columns: bool = True,
+):
+    """Generate markdown table
+
+    It will measure all the row cells and their header to generate
+    a human friendly table (the text representation looks like a table).
+
+    To avoid too much resizing when the cells change, the sizes are rounded
+    using `size_rounding`.
+
+    Example:
+    >>> print(create_table(
+    ...   ['key', 'value'],
+    ...   [['a', '1'], ['longer key here', '123456']],
+    ...   4))
+    | key              | value    |
+    |:-----------------|:---------|
+    | a                | 1        |
+    | longer key here  | 123456   |
+
+    Columns with values will be omitted by default:
+    >>> print(create_table(
+    ...   ['key', 'value', 'nothing'],
+    ...   [['a', '1', ''], ['longer key here', '123456', '']],
+    ...   4))
+    | key              | value    |
+    |:-----------------|:---------|
+    | a                | 1        |
+    | longer key here  | 123456   |
+
+    One can force them to be displayed regardless:
+    >>> print(create_table(
+    ...   ['key', 'value', 'nothing'],
+    ...   [['a', '1', ''], ['longer key here', '123456', '']],
+    ...   4, False))
+    | key              | value    | nothing  |
+    |:-----------------|:---------|:---------|
+    | a                | 1        |          |
+    | longer key here  | 123456   |          |
+    """
+    measure = create_rounded_str_len(size_rounding)
+    max_lengths = get_max_columns_lengths(
+        rows,
+        len(header),
+        measure,
+    )
+    if not omit_empty_columns:
+        def filter_visible_columns(seq):
+            return tuple(seq)
+    else:
+        visible_columns = [bool(ml) for ml in max_lengths]
+        def filter_visible_columns(seq):
+            return tuple(x for i, x in enumerate(seq) if visible_columns[i])
+
+    max_lengths = get_max_values(
+        max_lengths,
+        measure_items(header, measure),
+    )
+
+    visible_columns_lengths = filter_visible_columns(max_lengths)
+    fmt = create_table_format(visible_columns_lengths)
+    ret = [
+        fmt % filter_visible_columns(header),
+        create_table_header_separator(visible_columns_lengths),
+    ]
+    for r in rows:
+        ret.append(fmt % filter_visible_columns(r))
+    return "\n".join(ret)
+
+
+class DocumentEnum:
```

Comment:
I'd recommend to check for Markdown implementations first, before we implement this ourselves. Maybe https://github.com/Python-Markdown/markdown is worth a shot.

Reply:
not sure what you mean, I'm just generating https://raw.githubusercontent.com/GENIVI/vehicle_signal_specification/master/docs-gen/content/rule_set/data_entry/data_unit_types.md so it can be in sync with code

like the other comment, it's from that same file... my idea is just to write it in one place (ie: code) and have the documentation to be generated from it.

Reply:
I figured that from the code, but if we check the underlying issue this is not solving it? This makes it even harder to add units and the documentation we add is to brief to add any value imho. Easy fix would be: Update the .md file and be done with it. Why wouldn't we do that?

In addition, you wrote a markdown generator. Why reinvent the wheel, if we can reuse existing markdown libs?

Reply:
Because currently they diverged, see #90. The current process that needs to add documentation to one repo, code to another leads to divergences.

My proposal is to keep it in a single place (code) and provide a tool that generates the documentation. So https://github.com/GENIVI/vehicle_signal_specification tooling can be updated to execute this script *before* the markdown is converted to HTML to be used in the site.

As for the markdown generator, I'm just formatting the tables like they were in the https://raw.githubusercontent.com/GENIVI/vehicle_signal_specification/master/docs-gen/content/rule_set/data_entry/data_unit_types.md I didn't find any generator for that, do you? The repository you sent is something else, it **reads** markdown and outputs html. Note that this can be skipped, if nobody bothers to have a human readable (aligned) table, we can just output it in a simpler way that markdown will still understand and correctly format the HTML table.

Reply:
Ok, so your proposal is to generate markdown to afterwards manually add it to the wiki pages? What's the benifit here? The steps are essentially the same: You add/remove/change a unit, you run the tools, you update the documentation (which is not copy/paste instead of edit which may safe you 5 mins of work).

However, with this approach if you change the format (e.g. add/remove a column) you now have to do code changes to be in sync, whereas in the "classical" approach you just edit the markdown file.

Imho, we add  complexitiy to the tools here that make documentation more complex to change and instead of simplifing adding units as initially requested in https://github.com/GENIVI/vss-tools/issues/76 we make situation worse.

I understand, that those units and the docs should be in sync. Fair point. Let's just update the documentation and be done with it?

In addition, we have a setup here that we'd like to contribute where you can locally change the units config for your needs, which will help prototyping (as requested in https://github.com/GENIVI/vss-tools/issues/76). With the changes you propose here, this will get more complicated just to make a markdown table copyable. I think we should not add this complexitity.

Reply:
I don't know why you consider it manual. Since the CI scripts already do checkout the `vss-tools` code, it's a matter or running another command to output the markdown file in addition to to convert it to HTML. It needs automation, but doesn't require manual intervention, is that your understanding?

As for adding a new column bringing more work, it should be the same, no? You'd need to add the information to the markdown or Python code. You'd need to add a need to add the table header in markdown, so you do in this class (add a string to `table_header` tuple below. Shall you want to avoid adding the new column to all of the entries (you'd have to do that in markdown as well), you can define a default value in `VSSConstant` or right in the loop below (that calls`rows.append`), which you can't do in markdown. That said, if you could share more details on your Point of View, why it would be more work, let me know.

I can't comment on the code I didn't see, but the registry or file format can incorporate the documentation in a similar way. Say you have a constant name and a value, you can add description and domain as well. Even if that's optional, no?

Reply:
It's manual, since there is not automation in place that does change the markdown file automatically.

I assume it's an order of magnitude more complex to change python code than the markdown itself and requires a different skillset.

Reply:
Interesting exchange of ideas here.  I'll try to be brief, but a lot to cover.  

I often like the simple approach of just changing the files (documentation) and move on like @floix suggests.  But we have seen at least one case of this failing.  There is a proposal here so we should look at it I think.  But there are some points to cover first:

1) Whether or not the project should allow for adding your own units is still an open question!   We're debating the pros and cons.   (And #76 is perhaps no longer really an appropriate ticket to track the discussions we are having on the weekly VSS development call, primarily because we want to treat datatypes and units separately).  #91 was created to clarify the current state of the conversation in the VSS development group.

2) I agree with the logic that this PR basically depends on **if** there should be extensible units in the first place.
But see below for details on how to resolve that first.  We should recognize that it is also possible to remove code, if a strategy is later changed.  In some cases the right thing to do is to fix things for the _current_ situation.

The rest is best answered by quoting:

@floix wrote:
> It's manual, since there is not automation in place that does change the markdown file automatically.

I don't understand why you keep pushing this since @barbieri explained the plan for CI to do this?  Actually, GitHub actions already rebuild the docs for every change on master branch for a few weeks now, ref: [this line in the workflow](https://github.com/GENIVI/vehicle_signal_specification/blob/049b53fefd1f0bccc4a9f431a893cc03ecd10ae2/.github/workflows/buildcheck.yml#L46), and it could take care of running this tool too.  If you here only mean that this PR does not **yet** include the change to add that to the GitHub Actions workflow, then that seems an unnecessary nitpick.  (We actually still have the complication in VSS development that the Makefile is currently owned by **vehicle_signal_specification** repository instead of **vss-tools** repository, so it is not really possible to include the change in the same PR anyhow).

> I assume it's an order of magnitude more complex to change python code than the markdown itself and requires a different skillset.

Reasonable point although "order of magnitude" is a subjective evaluation.  But the alternative to this PR seems to be "change _both_ **code** and **docs**, or change only **code**", so does it really apply_

Basically we seem to have these possible futures:

A) A mechanism is added to add your own custom units by a separate configuration file (no "code" change)
or:
B) Possible units are specified in the vss-tools python code, as today:
    B1) Units are specified in code, and have to also be added to the markdown documentation at the same time.
    B2) Documentation is generated from the specification of units in the code file.  (=this PR)

@floix:
> In addition, we have a setup here that we'd like to contribute 
...

In such situations, I think it's fine to say:  Please wait with this change a week or two, because we have another idea coming.  But it cannot hold up the progress for a very long time  - then you have to present the idea at least (even if the code is taking a bit longer).  And in this case here, it is the fundamental question (#76, #91) that need to be dealt with primarily.

@floix:
> where you can locally change the units config for your needs, which will help prototyping (as requested in #76).

I would note that "we have something we'd like to contribute" is just more along the same lines as some other PRs that we found were coming too early**.  I feel that vss-tools is struggling from code change proposals to vss-tools being used to push some fairly major fundamental policy changes for VSS development (as I see it).  I think we see that it is not a good way -- this is why we raise issues and discuss the really fundamental things separately.  Then the PRs should just reflect the necessary code changes for the decision.  (And while I'm on the subject I have to say something I kept quiet about last week:  #91 is a VSS question first, before it becomes a vss-tools question, so the issue should have been raised in the VSS repository and not vss-tools.  *Because* these things are complex and have long lasting effects, I think we should be careful about the process.)

(** I think there was consensus on the VSS call that @barbieri could propose this PR so it could be evaluated.)

Reply:
Minor answer also to this (it is less about the big picture and more about details)

@Floix:
> In addition, you wrote a markdown generator. Why reinvent the wheel, if we can reuse existing markdown libs?

My view.  It is not _always_ better to add additional dependencies, if the code is trivial and can be written without adding  dependencies to something that we only use a tiny part of.  But this has to be evaluated for each case.  So I would second @barbieri 's question whether there is a reasonable choice to use that can generate these types of tables easily, and if so let's compare with that choice...

Reply:
about the last point, also note I can shrink the code if there is no need to render properly aligned *textual* table. Markdown itself will allow tables that are not perfectly aligned, so it's much simpler to generate (no need to measure, create the formatting strings and then loop again to print the table).

However, since the code is not that big and being able to read a nicely aligned textual table is always helpful, I'd like to keep what I did.